### PR TITLE
feat(datasource/custom): add ability for local registry (file://)

### PR DESCRIPTION
**Pull request by nSimonFR**
## Context

Renovate already allows to use your own dependencies from any HTTP server => https://docs.renovatebot.com/modules/datasource/custom/#custom-offline-dependencies

But sometimes, you may not want to expose your dependencies, or those may be local from your current repository, thus, the need to use local files as registry.

This allows a setup as such:
- renovate can manage dependencies from "remote" providers
- and use a "local" provider - for example to enforce the alignment of dependencies between two similar files and a regex matcher / custom local datasource

My organization especially needs this to "buffer" dependencies between two different environments (Our usecase is kubernetes manifests for [pre-]prod on the same `main` branch), this will allow us to use renovate as an automated way to manage our envs - without exposing our services versions through the internet.

This can be seen as an alternative to #23410 - a simpler version without the limitation for self-hosting renovate - especially useful for usecases such as https://github.com/renovatebot/renovate/issues/23410#issuecomment-1655947633.

## Changes

- Refactor `json` format and move it to `format` folder
- Add a `read` method to all `format` files that parses through filesystem
- ~Extend `CustomDatasourceConfig`~ use `file://`
- Extend datasource tests
- Extend custom datasource docs

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/custom/formats/json.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import fs from 'fs-extra';
```
Comment (viceice) [inline lib/modules/datasource/custom/formats/json.ts:1]: use  utils/fs instead

File: lib/modules/datasource/custom/index.ts [hunk 2]
```diff
@@ -25,26 +26,45 @@ export class CustomDatasource extends Datasource {
...
+            data = await yamlFetch(this.http, defaultRegistryUrlTemplate);
+            break;
+          case 'json':
+            data = await jsonFetch(this.http, defaultRegistryUrlTemplate);
+            break;
+        }
+      } catch (e) {
+        this.handleHttpErrors(e);
+        return null;
       }
```
Comment (nSimonFR) [inline lib/modules/datasource/custom/index.ts:65]: Self-reviewing: maybe the if/else isn't ideal, I could move it inside each format as a parameter to simplify the logic in this controller - but unsure
Also, my first implementation was to keep one big switch and use ternary in cases but it would throw in the catch http block which isn't ideal

File: lib/modules/datasource/custom/formats/plain.ts [hunk 3]
```diff
@@ -1,23 +1,36 @@
+import fs from 'fs-extra';
 import type { Http } from '../../../../util/http';
 import { newlineRegex } from '../../../../util/regex';
 import type { ReleaseResult } from '../../types';
 
+const convertLinesToVersions = (content: string): ReleaseResult | null => {
```
Comment (viceice) [inline lib/modules/datasource/custom/formats/plain.ts:6]: use normal function declaration and do refactor in separate PR

**Review by viceice — DISMISSED** [review]

File: lib/modules/datasource/custom/readme.md [hunk 4]
```diff
@@ -310,3 +310,21 @@ And the following custom manager:
...
 ```
+
+Or if you have the datasource locally, you can also use a local registry with the `isLocalRegistry` option:
+
+```json
+{
+  "customDatasources": {
+    "local_generic": {
+      "isLocalRegistry": true,
+      "defaultRegistryUrlTemplate": "dependencies/{{packageName}}/versiontracker.json",
```
Comment (nSimonFR) [inline lib/modules/datasource/custom/readme.md:321]: To be fair, I'm not 100% sure if this would be a correct API, maybe this shouldn't say `registryUrl` but `registryPath` ?

File: lib/config/types.ts [hunk 5]
```diff
@@ -264,6 +264,7 @@ export interface RenovateConfig
 
 export interface CustomDatasourceConfig {
   defaultRegistryUrlTemplate?: string;
+  isLocalRegistry?: boolean;
```
Comment (secustor) [inline lib/config/types.ts:267]:
```suggestion
```
Rather than `isLocalRegistry` use the file prefix `file://` in `registryUrl`s

**Review by secustor — CHANGES_REQUESTED** [review]

**Comment by viceice** [comment]
@rarkins what's your opinion about this idea?

**Comment by daveconde** [comment]
This would be very useful for the case of GitOps promotions where we'd like to be able to force specific versions are picked up.

**Comment by daveconde** [comment]
Is there anything I can do to help get this merged?

**Comment by rarkins** [comment]
Does this support reading both local files (within the repo) as well as absolute files off the operating system, or just within the repo?

If it's within the repo, what's the use case here? e.g. first PR updates some file1 and is merged, then a second PR updating file2 based off file1's contents is made using this capability? And in the meantime file1 and file2 are "out of sync" on the main branch?

**Comment by nSimonFR** [comment]
> Does this support reading both local files (within the repo) as well as absolute files off the operating system, or just within the repo?

I've never used renovate self-hosted, but the implementation uses the `readLocalFile` from lib and takes any path, so I would assume, both ?

> If it's within the repo, what's the use case here? e.g. first PR updates some file1 and is merged, then a second PR updating file2 based off file1's contents is made using this capability? And in the meantime file1 and file2 are "out of sync" on the main branch?

Yes, this is indeed the exact usecase I am trying to emulate, which would allow renovate to manage (and automate) our "pre-production" and "production" gitops manifests in the same branches.

_I thought about using a custom GHA for this but it seemed like extending renovate custom datasource was appropriate too._
If there are other ways to accomplish such a thing with `renovate`, I'd gladly take it too !

**Comment by rarkins** [comment]
I'm ok with this being for local files in the repo

**Comment by viceice** [comment]
`readLocalFile` only allows files from repo, it throws errors when trying to read outside of the repo.

**Review by secustor — APPROVED** [review]
I will look into moving this to a class based approach after the merge of this PR

**Review by rarkins — APPROVED** [review]
